File: scala/sdktest/UtilDirectTest.scala
// Generated direct-call tests for the util entity (mirrors the java
// TestDirect generator). A dependency-free scala-cli test object driven by
// SdkEntityTestMain: an offline mock transport records each call and the
// asserts confirm path-param substitution and the response shape.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.function.{BiFunction, Supplier}

import voxgig.thesmsworkssdk.core.{Helpers, ThesmsworksSDK}

object UtilDirectTest {

  private class DirectSetup(val client: ThesmsworksSDK, val calls: JList[JMap[String, Object]])

  private def directSetup(mockres: Object): DirectSetup = {
    val calls = new ArrayList[JMap[String, Object]]()
    val mockdata: Object = if (mockres != null) mockres else SdkTestSupport.om("id" -> "direct01")
    val mockFetch: BiFunction[String, JMap[String, Object], Object] =
      (url, init) => {
        calls.add(SdkTestSupport.om("url" -> url, "init" -> init))
        val js: Supplier[Object] = () => mockdata
        SdkTestSupport.om(
          "status" -> SdkTestSupport.I(200),
          "statusText" -> "OK",
          "headers" -> new LinkedHashMap[String, Object](),
          "json" -> js)
      }
    val client = new ThesmsworksSDK(SdkTestSupport.om(
      "base" -> "http://localhost:8080",
      "system" -> SdkTestSupport.om("fetch" -> mockFetch)))
    new DirectSetup(client, calls)
  }

  def run(rep: SdkTestReport): Unit = {
    rep.scope("direct-load-util") {
      val setup = directSetup(SdkTestSupport.om("id" -> "direct01"))
      val client = setup.client

      val params = new LinkedHashMap[String, Object]()
      params.put("errorcode", "direct01")
      val result = client.direct(SdkTestSupport.om(
          "path" -> "utils/errors/{errorcode}",
          "method" -> "GET",
          "params" -> params))

      rep.eq("direct-load-util.ok", java.lang.Boolean.TRUE, result.get("ok"))
      rep.eqI("direct-load-util.status", 200, Helpers.toInt(result.get("status")))
      rep.check("direct-load-util.data", result.get("data") != null, "expected data to be non-null")
      val dataMap = Helpers.toMapAny(result.get("data"))
      if (dataMap != null) rep.eq("direct-load-util.dataId", "direct01", dataMap.get("id"))
      rep.eqI("direct-load-util.calls", 1, setup.calls.size())
      val url = setup.calls.get(0).get("url") match { case s: String => s; case _ => "" }
      rep.check("direct-load-util.url1", url.contains("direct01"), "expected url to contain direct01, got " + url)
    }
  }
}

File: c/tests/util_direct_test.c
// Generated direct-call test for the util entity (mirrors the
// rust TestDirect generator; unit mode uses a mock system.fetch transport).

#include "ctest.h"

#include <stdlib.h>
#include <string.h>

static int CALLS = 0;
static char LAST_URL[1024];

// Mock transport: records the call, returns 200 + the ud response as json.
static voxgig_value* util_mock(void* ud, voxgig_value* args) {
  CALLS++;
  voxgig_value* url = voxgig_getelem(args, v_int(0), NULL);
  if (voxgig_is_string(url)) {
    snprintf(LAST_URL, sizeof(LAST_URL), "%s", voxgig_as_string(url));
  }
  voxgig_value* data = (voxgig_value*)ud;
  return cmap(4,
    "status", v_num(200),
    "statusText", v_str("OK"),
    "headers", v_map(),
    "json", json_thunk(data));
}

static ThesmsworksSDK* util_direct_setup(voxgig_value* mockres) {
  voxgig_value* opts = cmap(2,
    "base", v_str("http://localhost:8080"),
    "system", cmap(1, "fetch", vfn(util_mock, mockres)));
  return thesmsworks_sdk_new(opts);
}

int main(void) {

  // LOAD
  {
    CALLS = 0;
    voxgig_value* mockres = cmap(1, "id", v_str("direct01"));
    ThesmsworksSDK* sdk = util_direct_setup(mockres);
    voxgig_value* params = v_map();
    setp(params, "errorcode", v_str("direct01"));
    PNError* err = NULL;
    voxgig_value* result = sdk_direct(sdk, cmap(3,
      "path", v_str("utils/errors/{errorcode}"),
      "method", v_str("GET"),
      "params", params), &err);
    CHECK(err == NULL, "load: no error");
    voxgig_value* okv = getp(result, "ok");
    CHECK(voxgig_is_bool(okv) && voxgig_as_bool(okv), "load: ok true");
    CHECK_INT_EQ(to_int(getp(result, "status")), 200, "load: status 200");
    voxgig_value* data = getp(result, "data");
    CHECK(!v_is_noval(data), "load: data present");
    if (voxgig_is_map(data)) {
      CHECK_STR_EQ(get_str(data, "id"), "direct01", "load: data.id");
    }
    CHECK_INT_EQ(CALLS, 1, "load: one call");
    CHECK(strstr(LAST_URL, "direct01") != NULL, "load: url has direct01");
  }

  TEST_SUMMARY("thesmsworks_util_direct");
}
